src/backend/linux_raw/fs/raw_dir.rs

```diff
@@ -0,0 +1,201 @@
+use core::mem::MaybeUninit;
+use core::slice;
+use linux_raw_sys::general::linux_dirent64;
+
+use crate::backend::fs::syscalls::getdents_uninit;
+use crate::fd::AsFd;
+use crate::ffi::CStr;
+use crate::fs::FileType;
+use crate::io;
+
+/// A directory iterator implemented with getdents.
+///
+/// Note: This implementation does not handle growing the buffer. If this functionality is
+/// necessary, you'll need to drop the current iterator, resize the buffer, and then
+/// re-create the iterator. The iterator is guaranteed to continue where it left off provided
+/// the file descriptor isn't changed. See the example in [`RawDir::new`].
+pub struct RawDir<'buf, Fd: AsFd> {
+    fd: Fd,
+    buf: &'buf mut [MaybeUninit<u8>],
+    initialized: usize,
+    offset: usize,
+}
+
+impl<'buf, Fd: AsFd> RawDir<'buf, Fd> {
+    /// Create a new iterator from the given file descriptor and buffer.
+    ///
+    /// # Examples
+    ///
+    /// Using the heap:
+    ///
+    /// ```
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// let mut buf = Vec::with_capacity(8192);
+    /// for entry in RawDir::new(fd, buf.spare_capacity_mut()) {
+    ///     let entry = entry.unwrap();
+    ///     dbg!(&entry);
+    /// }
+    /// ```
+    ///
+    /// Using the stack:
+    ///
+    /// ```
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// let mut buf = [MaybeUninit::uninit(); 2048];
+    /// for entry in RawDir::new(fd, &mut buf) {
+    ///     let entry = entry.unwrap();
+    ///     dbg!(&entry);
+    /// }
+    /// ```
+    ///
+    /// When reading directory entries with arbitrarily large file names is required:
+    ///
+    /// ```
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    /// # use rustix::io::Errno;
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// let mut buf = Vec::with_capacity(8192);
+    /// 'read: loop {
+    ///     'resize: {
+    ///         for entry in RawDir::new(&fd, buf.spare_capacity_mut()) {
+    ///             let entry = match entry {
+    ///                 Err(Errno::INVAL) => break 'resize,
+    ///                 r => r.unwrap(),
+    ///             };
+    ///             dbg!(&entry);
+    ///         }
+    ///         break 'read;
+    ///     }
+    ///
+    ///     let new_capacity = buf.capacity() * 2;
+    ///     buf.reserve(new_capacity);
+    /// }
+    /// ```
+    pub fn new(fd: Fd, buf: &'buf mut [MaybeUninit<u8>]) -> Self {
```

Comment:
Do you know if `getdents` requires the buffer to be aligned at all? If so, we should document that, and either change the type here or potentially make this `unsafe`.

Reply:
Based on some experimentation, no:

```rust
        slice::from_raw_parts_mut(
            buf.as_mut_ptr().add(3) as *mut MaybeUninit<u8>,
            buf.capacity() - 3,
        )
```

That still works.

And this popped up in a search: https://github.com/microsoft/WSL/issues/1769

> Anyway it **isn't** an ABI requirement the pointer be 8 byte aligned

So I'm going to say I'm fairly confident the buffer doesn't have to be aligned. I would also expect the syscall to fail if it needs to be aligned which means it isn't a safety issue.